File: app/static/js/lib/device-update.js
/** Pure helpers for device software-update banner copy/state. */

export const LICENSE_UPDATE_REASONS = {
  EXPIRED: 'your license has expired',
  UNLICENSED:
    'you need to attach a valid TinyPilot Pro license to this device to retrieve updates',
  INVALID: 'there is an issue with your license',
};

export function licenseReason(licenseCheckStatus) {
  return (
    LICENSE_UPDATE_REASONS[licenseCheckStatus] || LICENSE_UPDATE_REASONS.INVALID
  );
}

/**
 * @returns {'hidden'|'updating'|'available'|'license-blocked'|'error'|null}
 */
export function updateBannerKind(softwareUpdate, currentVersion) {
  if (!softwareUpdate) {
    return null;
  }
  const jobStatus = softwareUpdate.job?.status;
  if (jobStatus === 'IN_PROGRESS') {
    return 'updating';
  }
  if (softwareUpdate.latest_error && !softwareUpdate.latest) {
    return 'error';
  }
  if (!softwareUpdate.update_available) {
    return 'hidden';
  }
  if (softwareUpdate.can_start) {
    return 'available';
  }
  return 'license-blocked';
}

export function latestVersionFrom(softwareUpdate) {
  return softwareUpdate?.latest?.version || '';
}

export function licenseStatusFrom(softwareUpdate) {
  return softwareUpdate?.latest?.licenseCheckStatus || '';
}

File: app/device_update.py
"""Helpers for TinyPilot device software-update banners."""


def license_blocks_update(license_check_status: str | None) -> bool:
    """Return True when Gatekeeper/Web UI would refuse an update kickoff."""
    if not license_check_status:
        return True
    return license_check_status != 'VALID'


def update_available(current_version: str | None, latest_version: str | None) -> bool:
    if not current_version or not latest_version:
        return False
    return current_version != latest_version


def can_start_update(
    current_version: str | None,
    latest_version: str | None,
    license_check_status: str | None,
) -> bool:
    return update_available(current_version, latest_version) and not license_blocks_update(
        license_check_status
    )
